File: src/algorithms/PalindromePartitioning.java
package algorithms; 

import java.util.*; 

public class PalindromePartitioning {
    /* Function: Return all the partitions of a string that all parts are palindrome. 
     * Solution: DFS, use a matrix to preprocess the palindrome judgement. 
    */
    public List<List<String>> partition(String s) {
        List<List<String>> res = new ArrayList<List<String>>(); 
        if (s == null || s.length() == 0) {
            return res; 
        }
        
        boolean[][] matrix = new boolean[s.length()][s.length()]; 
        List<String> list = new ArrayList<String>(); 
        partitionHelper(s, matrix, 0, list, res); 
        
        return res; 
    }
    
    private void partitionHelper(String s, boolean[][] matrix, int st, List<String> list, List<List<String>> res) {
        if (st == s.length()) {
            res.add(new ArrayList<String>(list)); 
            return; 
        }
        
        for (int i = st; i < s.length(); ++i) {
            if (isPalindrome(s, st, i, matrix)) {
                list.add(s.substring(st, i + 1)); 
                partitionHelper(s, matrix, i + 1, list, res); 
                list.remove(list.size() - 1); 
            }
        }
    }
    
    private boolean isPalindrome(String s, int st, int ed, boolean[][] matrix) {
        if (s.charAt(st) == s.charAt(ed) && (st + 1 >= ed || matrix[st + 1][ed - 1])) {
            matrix[st][ed] = true; 
            return true; 
        }
        
        return false; 
    }
}

File: src/CPP/src/leetcode/PalindromePartitioning.cpp
/*
 * [Source] https://leetcode.com/problems/palindrome-partitioning/
 * [Difficulty]: Medium
 * [Tag]: Backtracking
 */

#include <iostream>

using namespace std;

// [Solution]: start from 0 to end, if s[0,i] is palindrome, then add s[0,i] and recursive call to deal with s[i+1,end]. Use s[0]==s[i] && isPalindrome(s[1,i-1]) to fast the process.
// [Corner Case]:
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/rcanddp/PalindromePartitioning.java
 */

/* Java solution
public class Solution {
    public ArrayList<ArrayList<String>> partition(String s) {
		ArrayList<String> prev=new ArrayList<String>();
		ArrayList<ArrayList<String>> result=new ArrayList<ArrayList<String>>();
		partition(s, 0, prev, result);
		return result;
    }
	
	public void partition(String s, int pos, ArrayList<String> prev, ArrayList<ArrayList<String>> result){
		if(pos==s.length()) return;
		partition(s, pos+1, prev, result);
		if(!isPalindrome(s, pos)) return;
		ArrayList<String> addition=(ArrayList<String>)prev.clone();
		addition.add(s.substring(0, pos+1));
		if(pos==s.length()-1){
			result.add(addition);
			return;
		}
		partition(s.substring(pos+1), 0, addition, result);
	}
	
	public boolean isPalindrome(String s, int end){
		int start=0; 
		while(start<end){
			if(s.charAt(start)!=s.charAt(end)) return false;
			start++; end--;
		}
		return true;
	}
}
 */

int main() {
    Solution sol;

    return 0;
}
